src/m365/spo/commands/tenant/tenant-commandset-add.ts

```diff
@@ -0,0 +1,249 @@
+import { Logger } from '../../../../cli/Logger';
+import GlobalOptions from '../../../../GlobalOptions';
+import { validation } from '../../../../utils/validation';
+import SpoCommand from '../../../base/SpoCommand';
+import commands from '../../commands';
+import Command from '../../../../Command';
+import { Cli, CommandOutput } from '../../../../cli/Cli';
+import { Options as spoListItemAddCommandOptions } from '../listitem/listitem-add';
+import { Options as spoListItemListCommandOptions } from '../listitem/listitem-list';
+import * as spoTenantAppCatalogUrlGetCommand from './tenant-appcatalogurl-get';
+import * as spoListItemAddCommand from '../listitem/listitem-add';
+import * as spoListItemListCommand from '../listitem/listitem-list';
+import { urlUtil } from '../../../../utils/urlUtil';
+import { Solution } from './Solution';
+
+interface CommandArgs {
+  options: Options;
+}
+
+interface Options extends GlobalOptions {
+  title: string;
+  listType: string;
+  clientSideComponentId: string;
+  clientSideComponentProperties?: string;
+  webTemplate?: string;
+  location?: string;
+}
+
+class SpoTenantCommandSetAddCommand extends SpoCommand {
+  private static readonly listTypes: string[] = ['List', 'Library', 'SitePages'];
+  private static readonly locations: string[] = ['ContextMenu', 'CommandBar', 'Both'];
+
+  public get name(): string {
+    return commands.TENANT_COMMANDSET_ADD;
+  }
+
+  public get description(): string {
+    return 'Add a ListView Command Set as a tenant wide extension.';
+  }
+
+  constructor() {
+    super();
+
+    this.#initTelemetry();
+    this.#initOptions();
+    this.#initValidators();
+  }
+
+  #initTelemetry(): void {
+    this.telemetry.push((args: CommandArgs) => {
+      Object.assign(this.telemetryProperties, {
+        listType: args.options.listType,
+        clientSideComponentProperties: typeof args.options.clientSideComponentProperties !== 'undefined',
+        webTemplate: typeof args.options.webTemplate !== 'undefined',
+        location: args.options.location
+      });
+    });
+  }
+
+  #initOptions(): void {
+    this.options.unshift(
+      {
+        option: '-t, --title <title>'
+      },
+      {
+        option: '-l, --listType <listType>'
+      },
+      {
+        option: '-i, --clientSideComponentId <clientSideComponentId>'
+      },
+      {
+        option: '-p, --clientSideComponentProperties [clientSideComponentProperties]'
+      },
+      {
+        option: '-w, --webTemplate [webTemplate]'
+      },
+      {
+        option: '--location [location]'
+      }
+    );
+  }
+
+  #initValidators(): void {
+    this.validators.push(
+      async (args: CommandArgs) => {
+        if (!validation.isValidGuid(args.options.clientSideComponentId)) {
+          return `${args.options.clientSideComponentId} is not a valid GUID`;
+        }
+
+        if (SpoTenantCommandSetAddCommand.listTypes.indexOf(args.options.listType) < 0) {
```

Comment:
if the `listType` optional (or at least should be optional) we should validate it only when it is specified right?

Reply:
The listType is required...